[GLIB] Don't require GApplication for bwrap sandbox

TingPing opened the pull request:
#### 2bab2a58e316087dbe1bb5aa601f5e57d4384801
<pre>
[GLIB] Don&apos;t require GApplication for bwrap sandbox
<a href="https://bugs.webkit.org/show_bug.cgi?id=255249">https://bugs.webkit.org/show_bug.cgi?id=255249</a>

Reviewed by Adrian Perez de Castro and Carlos Garcia Campos.

The app-id isn&apos;t that important to our usage in our bubblewrap sandbox
so instead of erroring we can generate an id without any real downsides.

* Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
(WebKit::applicationId):
(WebKit::createFlatpakInfo):

Canonical link: <a href="https://commits.webkit.org/262947@main">https://commits.webkit.org/262947@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/49d2edfe2132b58368cc491ecc7c9d2ed05ab0bf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2746 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2745 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2892 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4155 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3109 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2890 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2844 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2415 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2772 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3203 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2457 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3925 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/697 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2442 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2538 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2453 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2493 "10 flakes 5 failures") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3677 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2826 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2270 "5 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2489 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2445 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2482 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/364 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2639 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shivamidow commented on Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
```diff
@@ -124,20 +125,24 @@ int argumentsToFileDescriptor(const Vector<CString>& args, const char* name)
 }
 
-static const char* applicationId(GError** error)
+static String applicationId()
```
/* Unofficial Review */
Suggestion: ensureApplicationId()? Because this function always returns an application id.

shivamidow commented on Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
```diff
@@ -124,20 +125,24 @@ int argumentsToFileDescriptor(const Vector<CString>& args, const char* name)
+    if (app) {
+        const char* appID = g_application_get_application_id(app);
+        if (appID)
+            return String::fromUTF8(appID);
```
/* Unofficial Review */
```
if (auto* app = g_application_get_default()) {
    if (const auto* appID = g_application_get_application_id(app))
        return String::fromUTF8(appID);
}
```

shivamidow commented on Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
```diff
@@ -124,20 +125,24 @@ int argumentsToFileDescriptor(const Vector<CString>& args, const char* name)
+
+    // There must be some id for xdg-desktop-portal to function.
+    // While xdg-desktop-portal does use this for permissions it's value
+    // doesn't matter much in our usage as we don't use any permissioned APIs.
```
/* Unofficial Review */
Suggestion:
We create a random ID for permission of xdg-desktop-portal, but the ID is not necessarily valid since we don't use permission-required APIs.

carlosgcampos commented on Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
```diff
@@ -124,20 +125,24 @@ int argumentsToFileDescriptor(const Vector<CString>& args, const char* name)
-    if (!app) {
-        g_set_error_literal(error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA, "GApplication is required.");
-        return nullptr;
+    if (app) {
```
if (auto* app = g_application_get_default()) {

carlosgcampos commented on Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
```diff
@@ -124,20 +125,24 @@ int argumentsToFileDescriptor(const Vector<CString>& args, const char* name)
-        return nullptr;
+    if (app) {
+        const char* appID = g_application_get_application_id(app);
+        if (appID)
```
```suggestion
        if (const char* appID = g_application_get_application_id(app))
```

shivamidow commented on Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
```diff
@@ -146,15 +151,9 @@ static int createFlatpakInfo()
+        String appID = applicationId();
         GUniquePtr<GKeyFile> keyFile(g_key_file_new());
-        g_key_file_set_string(keyFile.get(), "Application", "name", appID);
+        g_key_file_set_string(keyFile.get(), "Application", "name", appID.utf8().data());
```
/* Unofficial Review */
appID.utf8().data() -> ensureApplicationID().utf8().data(). Then, we can remove the declaration of appID above.

By the way, the only user of the applicationID() is here. char* -> String -> String (copy) ->  char* is inefficient. I suggest keeping const char* instead of String for the return value.

TingPing replied:
The API is no longer `const char*` because it dynamically generates a string, so a copy will always be around. `String` vs `GUniquePtr<char>` isn't a big difference IMO.

shivamidow reviewed (commented):
Not sure of the necessity of this change. But this is an unofficial review, so I defer the judgement to formal reviewers.

mcatanzaro commented on Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
```diff
@@ -124,20 +125,24 @@ int argumentsToFileDescriptor(const Vector<CString>& args, const char* name)
+        return String::fromUTF8(programName);
+
+    // There must be some id for xdg-desktop-portal to function.
+    // While xdg-desktop-portal does use this for permissions it's value
```
```suggestion
    // While xdg-desktop-portal does use this for permissions its value
```

mcatanzaro commented on Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
```diff
@@ -124,20 +125,24 @@ int argumentsToFileDescriptor(const Vector<CString>& args, const char* name)
+
+    // There must be some id for xdg-desktop-portal to function.
+    // While xdg-desktop-portal does use this for permissions it's value
+    // doesn't matter much in our usage as we don't use any permissioned APIs.
```
I highly doubt we can actually guarantee this comment remains true going forward, so I would reword it to instead say the best we can do is use a new app ID with new permissions rather than say that we don't do anything that requires checking permissions.

However, I worry that by using a new UUID each time we may bloat the permissions store. It's probably designed with the expectation that there will be a small number of apps. We should probably store the UUID in the data directory to avoid this. Bonus: then it will be persistent, and permissions will work properly.

TingPing replied:
I agree with the wording.

I really don't want to start storing this string though. Nothing should be added to the permission store.

The types of features that have permissions are things that a web content process should never do. Like screen sharing or file choosers. These are always going to be handled by other processes in WebKit which does its own permissions.

mcatanzaro replied:
Feel fragile, because there's no way to enforce it. We won't notice if we start doing something that requires permissions, or if something we currently do that does not require a permission starts to do so in the future. The portal APIs don't have a way to assert that they'll never prompt for permission or to indicate that all permissions should be denied. But maybe we should add one....

TingPing replied:
I'm going to merge this for now and open an issue to track this idea if that's OK?

mcatanzaro replied:
I suppose, though I don't see what's so bad about storing the app ID. It's not exactly a big file.

aperezdc replied:
Well, storing the identifiers opens a new can of worms: When are the files deleted? What if writing to disk fails (out of disk space, or so)? What if files disappear between runs? And so on. I would rather forget about saving the identifiers for now and see how things go. We can always make a follow-up patch later if needed.

Alternatively, it could be better to open `/proc/self/exe`, calculate the SHA256 checksum of the file, and use it as stable identifier: as long as the program on disk doesn't change, the identifier will be the same without needing to store anything else on disk... But I digress, even that I would leave for a follow up. If needed.

mcatanzaro replied:
This might come back to bite us in a few years when our permission stores grow huge, though. Imagine if your cookies database permanently increased in size each time you launch the browser. Using a random app ID results in the same thing, except for portal permissions.

Proposed alternative: use one hardcoded ID so that all applications get the exact same fallback ID and all permissions are shared.

TingPing replied:
> Proposed alternative: use one hardcoded ID so that all applications get the exact same fallback ID and all permissions are shared.

If WebKitGTK happens to use a permissioned API, which is when this theoretical problem appears, then this becomes a security risk. If you grant Webcam/ScreenSharing/etc permissions to one application you grant it to all of them.

TingPing replied:
> Alternatively, it could be better to open /proc/self/exe, calculate the SHA256 checksum of the file, and use it as stable identifier: as long as the program on disk doesn't change, the identifier will be the same without needing to store anything else on disk

This sounds like a decent idea to me.

mcatanzaro replied:
I suppose. A new permission store entry every time the binary changes is much less bad than a new one every time you start the program.

TingPing replied:
https://github.com/WebKit/WebKit/pull/12761

TingPing committed:
[GLIB] Don't require GApplication for bwrap sandbox
https://bugs.webkit.org/show_bug.cgi?id=255249

Reviewed by Adrian Perez de Castro and Carlos Garcia Campos.

The app-id isn't that important to our usage in our bubblewrap sandbox
so instead of erroring we can generate an id without any real downsides.

* Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
(WebKit::applicationId):
(WebKit::createFlatpakInfo):

Canonical link: https://commits.webkit.org/262947@main
